### PR TITLE
[WIP] Agent: Added roles and modified validation

### PR DESCRIPTION
**Pull request by jvluso**

### CODE REVIEW

File: future-apps/actor/contracts/Actor.sol [hunk 1]
```diff
@@ -181,4 +178,28 @@ contract Actor is Vault, IERC165, IERC1271, IForwarder {
     function getSig(bytes data) internal pure returns (bytes4 sig) {
         assembly { sig := add(data, 0x20) }
     }
+
+    function addressAt(bytes data) internal pure returns (address) {
+        uint160 m = 0;
+        uint160 b = 0;
+
+        for (uint8 i = 0; i < 20; i++) {
```
Comment (izqui) [inline future-apps/actor/contracts/Actor.sol]: This is very inefficient, you should just do:

```
uint160 addr;
assembly { addr := mload(data, 0x20) }
```

File: future-apps/actor/contracts/Actor.sol [hunk 2]
```diff
@@ -181,4 +178,28 @@ contract Actor is Vault, IERC165, IERC1271, IForwarder {
...
+        }
+
+        return address(m);
+    }
+
+    function bytesAfter(bytes data, uint offset) internal pure returns (bytes) {
+        uint length = data.length - offset;
+        bytes memory result = new bytes(length);
+
+        for (uint8 i = 0; i < length; i++) {
```
Comment (izqui) [inline future-apps/actor/contracts/Actor.sol]: This is terribly inefficient, you should use `memcpy`: https://github.com/Arachnid/solidity-stringutils/blob/master/src/strings.sol#L45

File: future-apps/actor/contracts/Actor.sol [hunk 3]
```diff
@@ -114,19 +102,28 @@ contract Actor is Vault, IERC165, IERC1271, IForwarder {
             return true;
         }
 
+        address designatedSigner = addressAt(signature);
+        bytes memory designatedSignature = bytesAfter(signature,20);
+        //should return the entire signature except the address
+
+        uint256[] memory params = new uint256[](0);
+        if (!canPerform(designatedSigner,PRESIGN_HASH_ROLE,params)) {
```
Comment (izqui) [inline future-apps/actor/contracts/Actor.sol]:
```suggestion
        if (!canPerform(designatedSigner, PRESIGN_HASH_ROLE, params)) {
```

File: future-apps/actor/contracts/Actor.sol [hunk 4]
```diff
@@ -17,7 +17,6 @@ contract Actor is Vault, IERC165, IERC1271, IForwarder {
     bytes32 public constant EXECUTE_ROLE = keccak256("EXECUTE_ROLE");
     bytes32 public constant RUN_SCRIPT_ROLE = keccak256("RUN_SCRIPT_ROLE");
     bytes32 public constant PRESIGN_HASH_ROLE = keccak256("PRESIGN_HASH_ROLE");
```
Comment (izqui) [inline future-apps/actor/contracts/Actor.sol]: Given this change, we need a better name for the role

Comment (jvluso) [reply future-apps/actor/contracts/Actor.sol]: I can't really think of a better name. What do you think it should be? It's still presigning hashes.

**Comment by sohkai** [comment]
@jvluso You'll need to run `npm run bootstrap` from the root dir again.

**Comment by sohkai** [comment]
Closing due to inactivity.

**Pull request closed without merging**